python/ray/data/tests/test_execution_optimizer.py

```diff
@@ -259,6 +263,115 @@ def test_read_map_batches_operator_fusion(ray_start_regular_shared, enable_optim
+        self.block_size = block_size
+
+    def create_reader(self, **read_args):
+        return BlockReader(self.num_blocks, self.block_size)
+
+
+class BlockReader(Reader):
+    def __init__(self, num_blocks: int, block_size: int):
+        self.num_blocks = num_blocks
+        self.block_size = block_size
+
+    def estimate_inmemory_data_size(self):
+        return None
+
+    def get_read_tasks(self, parallelism: int):
+        def _blocks_generator():
+            for _ in range(self.num_blocks):
+                yield [1] * self.block_size
+
+        return parallelism * [
+            ReadTask(
+                lambda: _blocks_generator(),
+                BlockMetadata(
+                    num_rows=self.num_blocks,
+                    size_bytes=self.num_blocks * self.block_size * 8,
+                    schema=None,
+                    input_files=None,
+                    exec_stats=None,
+                ),
+            )
+        ]
+
+
+def test_operator_fusion_zero_copy_blocks_to_batches_e2e(
+    ray_start_regular_shared, enable_optimizer
+):
+    # Test that fused Read and MapBatches streams read-level blocks directly to
+    # MapBatches UDF.
+    num_blocks = 10
+    block_size = 5
+
+    def check_batch(batch):
+        # Check that the batch is the same size as the input block size.
+        assert len(batch) == block_size
+        return batch
+
+    planner = Planner()
+    read_op = Read(BlockDatasource(num_blocks, block_size), parallelism=8)
+    map_op = MapBatches(
+        read_op,
+        check_batch,
+        batch_size=None,
+    )
+    logical_plan = LogicalPlan(map_op)
+    physical_plan = planner.plan(logical_plan)
+    physical_plan = PhysicalOptimizer().optimize(physical_plan)
+    op = physical_plan.dag
+
+    executor = BulkExecutor(ExecutionOptions())
+    outputs = executor.execute(op)
+    out_bundles = [ray.get([block for block, _ in bundle.blocks]) for bundle in outputs]
+
+    # Sanity check.
+    assert len(out_bundles) == 8
+
+
+def test_operator_fusion_zero_copy_batch_to_batch_e2e(
+    ray_start_regular_shared, enable_optimizer
+):
+    # Test that fused MapBatches ops streams doesn't materialize batches into large
+    # blocks between the MapBatches.
+    num_blocks = 10
+    block_size = 10
+    batch_size = 5
+
+    planner = Planner()
+    read_op = Read(BlockDatasource(num_blocks, block_size), parallelism=8)
+    map_op1 = MapBatches(
+        read_op,
+        lambda x: x,
+        batch_size=batch_size,
+    )
+
+    def check_batch(batch):
+        # Check that the batch is the same size as the batch size if the upstream map.
+        assert len(batch) == batch_size
+        return batch
+
+    map_op2 = MapBatches(
+        map_op1,
+        check_batch,
+        batch_size=None,
+    )
+    logical_plan = LogicalPlan(map_op2)
+    physical_plan = planner.plan(logical_plan)
+    physical_plan = PhysicalOptimizer().optimize(physical_plan)
+    op = physical_plan.dag
```

Comment:
Can we improve the testing / observability by emitting metrics on batch conversion operations and overheads? Then we can just assert the expected conversion in the metrics, e.g. `{"numpy_to_pandas_conversions": 5}` etc.

Reply:
Should we do this?